### PR TITLE
multi currency incentive rewards

### PR DESCRIPTION
**Pull request by ermalkaleci**
Stash

### CODE REVIEW

File: modules/incentives/src/lib.rs [hunk 1]
```diff
@@ -222,7 +237,7 @@ pub mod module {
 		PoolId<T::RelaychainAccountId>,
 		Twox64Concat,
 		T::AccountId,
-		Balance,
+		BTreeMap<CurrencyId, Balance>,
```
Comment (xlc) [inline modules/incentives/src/lib.rs:240]: need migration

**Review by xlc — COMMENTED** [review]
Also need to add the migration for orml-rewards to karura runtime

**Review by xlc — COMMENTED** [review]
Need migration for orml-rewards

**Review by wangjj9219 — COMMENTED** [review]
plan to use `HomaValidatorAllowance` to implement the multi-currency reward? But reward accumulation for `HomaValidatorAllowance` is not periodic.

The definition for `PoolId` is no longer appropriate. `PooId` implies the reward currency id(native and stable or other allowance) and accumulation method （periodly by fixed amount ，or the multiple of  the base amount and fixed rate, or transfer allowance non-periodicly)

Now, `orml_rewards` supports to record multi-currency rewards amount with single `PoolId`, 
And the pool share of `DexIncentives` is equal to `DexSaving` for the same trading pair. 
it's better to refactor `PoolId` to `DEX`, `Loans` and others, and need new storage 
`map Vec<(reward_currency_id, reward_amount_per_period)>` 
to record periodic fixed amount reward, and define new enum to imply base amount source, add another storage 
`map PoolId => Vec<(reward_currency_id, fixed_rate, BaseAmountSourceEnum)>` to record the fixed rate and the base amount source. Aggregate these configurations when accumulating periodic multi rewards.

@xlc What do you think? 
If need to do, more migrations are required, and I am afraid that testing and front-end modification cannot be completed before this Thursday’s runtime upgrade.



 with PoolId to support native incentives and other currency reward.

File: modules/incentives/src/lib.rs [hunk 2]
```diff
@@ -328,10 +295,39 @@ pub mod module {
 				0
 			}
 		}
+
+		fn on_runtime_upgrade() -> Weight {
```
Comment (xlc) [inline modules/incentives/src/lib.rs:299]: we don't want this anymore right

Comment (wangjj9219) [reply modules/incentives/src/lib.rs:299]: But other storages still need to be migrated on `on_runtime_upgrade`.
Do you mean to add calls to trigger manual migration for all storage like `migrate_share_and_withdrawn_reward` and `migrate_pending_rewards`?

Comment (xlc) [reply modules/incentives/src/lib.rs:299]: right. that's fine then.

**Review by xlc — APPROVED** [review]

**Pull request merged**